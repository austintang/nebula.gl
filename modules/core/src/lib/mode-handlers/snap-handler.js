// @flow

import turfArea from '@turf/area';
import turfBearing from '@turf/bearing';
import turfCenter from '@turf/center';
import turfDistance from '@turf/distance';
import turfEnvelope from '@turf/envelope';
import { point as turfPoint } from '@turf/helpers';
import { coordAll } from '@turf/meta';
import turfTransformTranslate from '@turf/transform-translate';

import type { Feature, FeatureCollection } from '../../geojson-types.js';
import { SortedList, SortType } from '../utils';
import { ModeHandler } from './mode-handler';

type SnapDetails = {
  snapDistance: number,
  selectedSnapPoint: Feature,
  nonSelectedSnapPoint: Feature,
  index?: number
};

export class SnapHandler extends ModeHandler {
  _renderSnapEditHandles: boolean;
  _selectedSnapPoint: Feature;
  _nonSelectedSnapPoint: Feature;
  _snapAssociations: number[][];

  constructor(featureCollection?: FeatureCollection) {
    super(featureCollection);
    this._snapAssociations = [];
  }

  renderSnapHandles() {
    this._renderSnapEditHandles = true;
  }

  hideSnapHandles() {
    this._renderSnapEditHandles = false;
  }

  getSnapEditHandleFromPoint(pointFeature: any) {
    const { geometry } = pointFeature || {};
    if (geometry) {
      return {
        position: geometry.coordinates,
        type: 'intermediate',
        // Dummy values for positionIndexes and featureIndex as these properties are not used
        positionIndexes: [-1],
        featureIndex: -1
      };
    }
    return null;
  }

  shouldRenderSnapHandles() {
    const { enablePolygonSnapping } = this.getModeConfig() || {};
    return enablePolygonSnapping && this.isSinglePolygonSelected() && this._renderSnapEditHandles;
  }

  getEditHandles() {
    if (this.shouldRenderSnapHandles()) {
      return [
        this.getSnapEditHandleFromPoint(this._selectedSnapPoint),
        this.getSnapEditHandleFromPoint(this._nonSelectedSnapPoint)
      ].filter(Boolean);
    }
    return [];
  }

  // Get the indexes of polygons closest to the selected feature. The numberToTrack
  // optional parameter dictates how many polygons to track e.g. a numberToTrack = 3
  // will result in the 3 closest polygon indexes being returned by this method. These
  // polygons will be candidates for snapping.
  getNearestPolygonIndexes(options: ?{ [key: string]: any }) {
    let arraySize = 1;
    if (options) {
      arraySize = options.numberToTrack;
    }
    const [selectedIndex] = this.getSelectedFeatureIndexes();
    const minDistanceArray = new SortedList({
      arraySize,
      getValueToCompareFn: val => val.distance,
      sortType: SortType.MIN
    });

    const allFeatures = this.getImmutableFeatureCollection().getObject().features;
    const selectedCenter = turfCenter(this.getSelectedFeature());

    for (let i = 0; i < allFeatures.length; i++) {
      const currentFeature = allFeatures[i];
      if (i !== selectedIndex && currentFeature.geometry.type === 'Polygon') {
        const currentFeatureCenter = turfCenter(currentFeature);
        const distance = turfDistance(currentFeatureCenter, selectedCenter);
        minDistanceArray.push({ distance, index: i });
      }
    }
    return minDistanceArray.toArray().map(details => details.index);
  }

  isSinglePolygonSelected() {
    const selectedIndexes = this.getSelectedFeatureIndexes();
    const selectedFeature = this.getSelectedFeature();
    if (selectedFeature) {
      return selectedIndexes.length === 1 && selectedFeature.geometry.type === 'Polygon';
    }
    return false;
  }

  // Scale the snapping strength directly proportional to the selected polygon's size
  getSnapStrengthModifier() {
    const selectedFeature = this.getSelectedFeature();
    if (selectedFeature) {
      // Using @turf/envelope helps normalize the area of different polygon shapes/types
      // (i.e. polygon with holes, ciruclar polygons, etc.)
      const envelope = turfEnvelope(selectedFeature);
      const area = turfArea(envelope.geometry);
      return Math.pow(area, 0.45) / 2000;
    }
    return 1;
  }

  hasSelectedBeenSnapped() {
    const [selectedIndex] = this.getSelectedFeatureIndexes();
    const snapPoints = this._snapAssociations[selectedIndex];
    return snapPoints ? snapPoints.length : false;
  }

  shouldPerformSnap() {
    const { enablePolygonSnapping } = this.getModeConfig() || {};
    return (
      enablePolygonSnapping && this.isSinglePolygonSelected() && !this.hasSelectedBeenSnapped()
    );
  }

  shouldPerformUnsnap() {
    return this.isSinglePolygonSelected() && this.hasSelectedBeenSnapped();
  }

  shouldPerformStandardModeAction() {
    return !this.isSinglePolygonSelected() || !this.hasSelectedBeenSnapped();
  }

  _getLocalMinimumDistanceSnap(selectedVertexes: any[], nonSelectedVertexes: any[]): ?SnapDetails {
    let minSnapDistance = Number.MAX_VALUE;
    let snapDetails;

    for (const selectedVertex of selectedVertexes) {
      for (const nonSelectedVertex of nonSelectedVertexes) {
        const vertexDistance = turfDistance(selectedVertex, nonSelectedVertex);
        if (vertexDistance < minSnapDistance) {
          minSnapDistance = vertexDistance;

          snapDetails = {
            snapDistance: vertexDistance,
            selectedSnapPoint: selectedVertex,
            nonSelectedSnapPoint: nonSelectedVertex
          };
        }
      }
    }
    return snapDetails;
  }

  // Calculates the closest vertexes from both the selected polygon and
  // the non-selected polygon candidates passed in through polygonIndexes
  // method paramter.
  getSnapDetailsFromCandidates(polygonIndexes: number[]) {
    const selectedPolyon = this.getSelectedFeature();
    const selectedVertexes = coordAll(selectedPolyon).map(turfPoint);

    let minSnapDistance = Number.MAX_VALUE;
    let snapDetails;

    const { features } = this.getFeatureCollection();

    for (const index of polygonIndexes) {
      const polygon = features[index];
      const nonSelectedVertexes = coordAll(polygon).map(turfPoint);

      const localMinDistanceSnap = this._getLocalMinimumDistanceSnap(
        selectedVertexes,
        nonSelectedVertexes
      );

      const { snapDistance } = localMinDistanceSnap || {};
      if (snapDistance < minSnapDistance) {
        minSnapDistance = snapDistance;
        snapDetails = { ...localMinDistanceSnap, index };
      }
    }

    const { selectedSnapPoint, nonSelectedSnapPoint } = snapDetails || {};
    // Cache snap points for edit handle rendering
    this._selectedSnapPoint = selectedSnapPoint;
    this._nonSelectedSnapPoint = nonSelectedSnapPoint;
    return snapDetails;
  }

  // Calculate the final post-snap position of the selected polygon
  calculateSnapMove(snapDetails: Object, snapStrength: number) {
    if (snapDetails) {
      const { nonSelectedSnapPoint, selectedSnapPoint, snapDistance, index } = snapDetails;
      const snapStrengthModifier = this.getSnapStrengthModifier();

      if (snapDistance <= snapStrength * snapStrengthModifier) {
        const [selectedIndex] = this.getSelectedFeatureIndexes();
        const selectedFeature = this.getSelectedFeature();

        const snapBearing = turfBearing(selectedSnapPoint, nonSelectedSnapPoint);
        const movedPolygon = turfTransformTranslate(selectedFeature, snapDistance, snapBearing);
        this.cacheSnapAssociates(index, selectedIndex);

        // Post snap, both _selectedSnapPoint and _nonSelectedSnapPoint will equal
        // to nonSelectedSnapPoint
        this._selectedSnapPoint = nonSelectedSnapPoint;
        this._nonSelectedSnapPoint = nonSelectedSnapPoint;

        return {
          movedPolygon,
          selectedIndex
        };
      }
    }
    return null;
  }

  // Cache the indexes of the polygons that have participated in a snap
  cacheSnapAssociates(index: number, partnerIndex: number) {
    // Cache index to partnerIndex association
    if (!this._snapAssociations[index]) {
      this._snapAssociations[index] = [];
    }
    this._snapAssociations[index].push(partnerIndex);

    // Cache partnerIndex to index association
    if (!this._snapAssociations[partnerIndex]) {
      this._snapAssociations[partnerIndex] = [];
    }
    this._snapAssociations[partnerIndex].push(index);
  }

  getSnapAssociates(index: number) {
    return this._snapAssociations[index] || [];
  }

  // Clear all snap associations from the specified index
  clearSnapAssociates(index: number) {
    const snapPartnerIndexes = this._snapAssociations[index];
    if (!snapPartnerIndexes || !snapPartnerIndexes.length) return;

    // Clear the specified index from all associated partners
    for (const partnerIndex of snapPartnerIndexes) {
      const associatesOfPartner = this._snapAssociations[partnerIndex];
      if (associatesOfPartner) {
        this._snapAssociations[partnerIndex] = associatesOfPartner.filter(i => i !== index);
      }
    }

    // Clear all associations from index
    if (this._snapAssociations[index]) {
      this._snapAssociations[index] = [];
    }
  }
}
