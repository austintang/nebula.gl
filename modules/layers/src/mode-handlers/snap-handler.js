// @flow

import turfArea from '@turf/area';
import turfBearing from '@turf/bearing';
import turfCenter from '@turf/center';
import turfDistance from '@turf/distance';
import turfEnvelope from '@turf/envelope';
import { point as turfPoint } from '@turf/helpers';
import turfTransformTranslate from '@turf/transform-translate';

import type { FeatureCollection, Position } from '../geojson-types.js';
import { SortedList, SortType } from '../utils';
import type { EditHandle } from './mode-handler';
import { ModeHandler, getEditHandlesForGeometry } from './mode-handler';

export class SnapHandler extends ModeHandler {
  _pickedHandle: ?EditHandle;
  _potentialNonPickedHandle: ?EditHandle;
  _potentialSnapDistance: number;
  _snapAssociations: number[][];

  constructor(featureCollection?: FeatureCollection) {
    super(featureCollection);
    this._snapAssociations = [];
  }

  setPickedHandle(handle: ?EditHandle) {
    this._pickedHandle = handle;
  }

  clearCachedHandles() {
    this._pickedHandle = null;
    this._potentialNonPickedHandle = null;
  }

  updatePickedHandlePosition(index: number, updatedGeometry: any) {
    if (this._pickedHandle) {
      const { positionIndexes, featureIndex } = this._pickedHandle;
      if (index >= 0 && featureIndex !== index) return;

      const { coordinates } = updatedGeometry;
      (this._pickedHandle || {}).position = positionIndexes.reduce(
        (a: any[], b: number) => a[b],
        coordinates
      );
    }
  }

  getFeatureFromHandle(handle: ?EditHandle): ?any {
    if (!handle) return null;
    const { featureIndex } = handle;
    const { features } = this.featureCollection.getObject();
    if (featureIndex >= features.length) {
      console.warn(`featureCollection features out of range ${featureIndex}`); // eslint-disable-line no-console,no-undef
      return null;
    }
    return features[featureIndex];
  }

  _getNonPickedIntermediateHandles(): EditHandle[] {
    const handles = [];
    const { features } = this.featureCollection.getObject();

    for (let i = 0; i < features.length; i++) {
      if (!this.getSelectedFeatureIndexes().includes(i) && i < features.length) {
        const { geometry } = features[i];
        handles.push(...getEditHandlesForGeometry(geometry, i, 'intermediate'));
      } else if (i >= features.length) {
        console.warn(`selectedFeatureIndexes out of range ${i}`); // eslint-disable-line no-console,no-undef
      }
    }
    return handles;
  }

  // If no snap handle has been picked, only display the edit handles of the
  // selected feature. If a snap handle has been picked, display said snap handle
  // along with all snappable points on all non-selected features.
  getEditHandles(picks?: Array<Object>, groundCoords?: Position): any[] {
    const { enablePolygonSnapping } = this.getModeConfig() || {};
    const handles = [];
    if (!enablePolygonSnapping) return handles;

    if (this._pickedHandle) {
      handles.push(...this._getNonPickedIntermediateHandles(), this._pickedHandle);

      const nonPickedFeature = this.getFeatureFromHandle(this._potentialNonPickedHandle);
      if (nonPickedFeature) {
        handles.push(this._potentialNonPickedHandle);
      }
      return handles;
    }

    const { features } = this.featureCollection.getObject();
    for (const index of this.getSelectedFeatureIndexes()) {
      if (index < features.length) {
        const { geometry } = features[index];
        handles.push(...getEditHandlesForGeometry(geometry, index, 'snap'));
      } else {
        console.warn(`selectedFeatureIndexes out of range ${index}`); // eslint-disable-line no-console,no-undef
      }
    }

    return handles;
  }

  // Get the indexes of polygons closest to the selected feature. The numberToTrack
  // optional parameter dictates how many polygons to track e.g. a numberToTrack = 3
  // will result in the 3 closest polygon indexes being returned by this method. These
  // polygons will be candidates for snapping.
  _getNearestPolygonIndexes(options: ?{ numberOfPolygonsToTrack: number }): any[] {
    let arraySize = 1;
    if (options) {
      arraySize = options.numberOfPolygonsToTrack;
    }
    const selectedIndexes = this.getSelectedFeatureIndexes();
    const minDistanceArray = new SortedList({
      arraySize,
      getValueToCompareFn: val => val.distance,
      sortType: SortType.MIN
    });

    const { features } = this.featureCollection.getObject();
    const selectedFeature = this.getFeatureFromHandle(this._pickedHandle);

    if (selectedFeature) {
      const selectedCenter = turfCenter(selectedFeature);

      for (let i = 0; i < features.length; i++) {
        const currentFeature = features[i];
        if (!selectedIndexes.includes(i)) {
          const currentFeatureCenter = turfCenter(currentFeature);
          const distance = turfDistance(currentFeatureCenter, selectedCenter);
          minDistanceArray.push({ distance, index: i });
        }
      }
    }
    return minDistanceArray.toArray().map(details => details.index);
  }

  // Scale the snapping strength directly proportional to the selected polygon's size
  getSnapStrengthModifier(): number {
    if (this._pickedHandle) {
      const selectedFeature = this.getFeatureFromHandle(this._pickedHandle);

      if (selectedFeature) {
        // Using @turf/envelope helps normalize the area of different polygon shapes/types
        // (i.e. polygon with holes, ciruclar polygons, etc.)
        const envelope = turfEnvelope(selectedFeature);
        const area = turfArea(envelope.geometry);
        return Math.pow(area, 0.45) / 2000;
      }
    }
    return 1;
  }

  _hasSelectedBeenSnapped(): boolean {
    const selectedIndexes = this.getSelectedFeatureIndexes();
    if (!selectedIndexes.length) return false;
    return selectedIndexes
      .map(index => this._snapAssociations[index])
      .some(assoc => assoc && assoc.length);
  }

  _areAllSnapAssociatesSelected() {
    const selectedIndexes = this.getSelectedFeatureIndexes();
    if (!selectedIndexes.length) {
      return true;
    }
    return selectedIndexes
      .map(index => this.getSnapAssociates(index))
      .every(indexes => indexes.every(index => selectedIndexes.includes(index)));
  }

  _isNonPickedAlreadySnappedToPicked() {
    if (this._pickedHandle && this._potentialNonPickedHandle) {
      const { featureIndex: nonPickedIndex } = this._potentialNonPickedHandle;

      const { featureIndex: pickedIndex } = this._pickedHandle || {};
      const snapAssociates = this.getSnapAssociates(pickedIndex);
      return snapAssociates.includes(nonPickedIndex);
    }
    return false;
  }

  shouldPerformSnap(): boolean {
    const { enablePolygonSnapping } = this.getModeConfig() || {};
    if (enablePolygonSnapping) {
      this._cacheClosestNonPickedHandle({ numberOfPolygonsToTrack: 100 });
      return !this._isNonPickedAlreadySnappedToPicked();
    }
    return false;
  }

  shouldPerformUnsnap(): boolean {
    return this._hasSelectedBeenSnapped() && !this._areAllSnapAssociatesSelected();
  }

  shouldPerformStandardModeAction(): boolean {
    return !this._hasSelectedBeenSnapped() || this._areAllSnapAssociatesSelected();
  }

  _cacheClosestNonPickedHandle(options?: { numberOfPolygonsToTrack: number }) {
    if (!this._pickedHandle) return;
    const nonPickedHandles = this._getNonPickedIntermediateHandles();

    // Calculate the closest handle from non-selected features to the picked handle
    let minDistance = Number.MAX_VALUE;
    let nearestNonPickedHandle;
    for (const handle of nonPickedHandles) {
      const distance = turfDistance(
        this._pickedHandle && this._pickedHandle.position,
        handle.position
      );

      if (distance < minDistance) {
        minDistance = distance;
        nearestNonPickedHandle = handle;
      }
    }

    // Store the closest non-picked handle and the distance from the picked handle
    if (nearestNonPickedHandle) {
      this._potentialNonPickedHandle = nearestNonPickedHandle;
    }
    this._potentialSnapDistance = minDistance;
  }

  // Cache the indexes of the polygons that have participated in a snap
  cacheSnapAssociates(index: number, partnerIndex: number) {
    // Cache index to partnerIndex association
    if (!this._snapAssociations[index]) {
      this._snapAssociations[index] = [];
    }
    this._snapAssociations[index].push(partnerIndex);

    // Cache partnerIndex to index association
    if (!this._snapAssociations[partnerIndex]) {
      this._snapAssociations[partnerIndex] = [];
    }
    this._snapAssociations[partnerIndex].push(index);
  }

  getSnapAssociates(index: number): number[] {
    return this._snapAssociations[index] || [];
  }

  // Clear all snap associations from the specified index
  clearSnapAssociates(index: number, associateIndex: number) {
    const snapPartnerIndexes = this._snapAssociations[index];
    if (!snapPartnerIndexes || !snapPartnerIndexes.length) return;

    const associatePartnerIndexes = this._snapAssociations[associateIndex];
    if (!associatePartnerIndexes || !associatePartnerIndexes.length) return;

    this._snapAssociations[index] = snapPartnerIndexes.filter(i => i !== associateIndex);
    this._snapAssociations[associateIndex] = associatePartnerIndexes.filter(i => i !== index);
  }

  calculateSnapIfWithinThreshold(
    snapStrength: number
  ): ?({ movedFeature: any, selectedIndex: number }[]) {
    if (this._pickedHandle) {
      const modifiedSnapStrength = snapStrength * this.getSnapStrengthModifier();

      if (this._potentialSnapDistance <= modifiedSnapStrength && this._potentialNonPickedHandle) {
        const selectedFeatures = this.getSelectedFeaturesAsFeatureCollection();
        const selectedIndexes = this.getSelectedFeatureIndexes();

        const { position: pickedPosition } = this._pickedHandle || {};
        const { position: nonPickedPosition, featureIndex: nonPickedFeatureIndex } =
          this._potentialNonPickedHandle || {};

        // Calculate the post-snap position of the selected feature(s)
        const pickedHandlePoint = turfPoint(pickedPosition);
        const potentialSnapPoint = turfPoint(nonPickedPosition);
        const snapBearing = turfBearing(pickedHandlePoint, potentialSnapPoint);

        const movedFeatures = turfTransformTranslate(
          selectedFeatures,
          this._potentialSnapDistance,
          snapBearing
        );

        const postSnapPositions = [];

        for (let i = 0; i < selectedIndexes.length; i++) {
          const selectedIndex = selectedIndexes[i];
          const movedFeature = movedFeatures.features[i];
          postSnapPositions.push({
            movedFeature,
            selectedIndex
          });
        }

        const { featureIndex: selectedIndex } = this._pickedHandle || {};
        this.cacheSnapAssociates(selectedIndex, nonPickedFeatureIndex);

        return postSnapPositions;
      }
    }

    return null;
  }
}
