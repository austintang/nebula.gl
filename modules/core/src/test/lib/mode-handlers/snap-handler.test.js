// @flow

import { point as turfPoint } from '@turf/helpers';
import type { FeatureCollection } from '../../../geojson-types.js';
import { SnapHandler } from '../../../lib/mode-handlers/snap-handler.js';

export function mockFeatureCollectionState(polygon: any) {
  const handler = new SnapHandler(polygon);
  handler.setSelectedFeatureIndexes([1]);
  return handler;
}

const polygonsForSnappingTests: FeatureCollection = {
  type: 'FeatureCollection',
  features: [
    {
      type: 'Feature',
      properties: {},
      geometry: {
        type: 'Polygon',
        coordinates: [
          [
            [-122.49485401280788, 37.987923255302974],
            [-122.44252582524939, 37.987923255302974],
            [-122.44252847887157, 37.93873205786406],
            [-122.49485666643005, 37.93873205786406],
            [-122.49485401280788, 37.987923255302974]
          ]
        ]
      }
    },
    {
      type: 'Feature',
      properties: {},
      geometry: {
        type: 'Polygon',
        coordinates: [
          [
            [-122.40814940182412, 37.97528325161274],
            [-122.37737022011595, 37.97528325161274],
            [-122.37737142575622, 37.952934704562644],
            [-122.40815060746439, 37.952934704562644],
            [-122.40814940182412, 37.97528325161274]
          ]
        ]
      }
    },
    {
      type: 'Feature',
      properties: {},
      geometry: {
        type: 'Point',
        coordinates: [-122.28103267622373, 37.98843664327903]
      }
    },
    {
      type: 'Feature',
      properties: {},
      geometry: {
        type: 'Polygon',
        coordinates: [
          [
            [-122.40908525092362, 37.85902221692065],
            [-122.34432261530361, 37.85902221692065],
            [-122.3442469760231, 37.8157115979288],
            [-122.40900961164311, 37.8157115979288],
            [-122.40908525092362, 37.85902221692065]
          ]
        ]
      }
    },
    {
      type: 'Feature',
      properties: {},
      geometry: {
        type: 'Polygon',
        coordinates: [
          [
            [-122.40682264287454, 38.13330760982133],
            [-122.34143228624993, 38.13330760982133],
            [-122.34144710122638, 38.08906337516829],
            [-122.40683745785105, 38.08906337516829],
            [-122.40682264287454, 38.13330760982133]
          ]
        ]
      }
    }
  ]
};

describe('polygon snapping', () => {
  let handler;

  beforeEach(() => {
    handler = mockFeatureCollectionState(polygonsForSnappingTests);
  });

  test('getSnapEditHandleFromPoint()', () => {
    const snapHandle = handler.getSnapEditHandleFromPoint(turfPoint([1, 2]));
    const expectedSnapHandle = {
      position: [1, 2],
      type: 'intermediate',
      featureIndex: -1,
      positionIndexes: [-1]
    };
    expect(snapHandle).toEqual(expectedSnapHandle);
  });

  test('shouldRenderSnapHandles() - many polygons selected', () => {
    handler.setSelectedFeatureIndexes([0, 1]);
    expect(handler.shouldRenderSnapHandles()).toBeFalsy();
  });

  test('shouldRenderSnapHandles() - point selected', () => {
    handler.setSelectedFeatureIndexes([2]);
    expect(handler.shouldRenderSnapHandles()).toBeFalsy();
  });

  test('shouldRenderSnapHandles() - no feature selected', () => {
    handler.setSelectedFeatureIndexes([]);
    expect(handler.shouldRenderSnapHandles()).toBeFalsy();
  });

  test('shouldRenderSnapHandles() - enablePolygonSnapping and _renderSnapEditHandles are false', () => {
    expect(handler.shouldRenderSnapHandles()).toBeFalsy();
  });

  test('shouldRenderSnapHandles() - positive case', () => {
    handler.setModeConfig({ enablePolygonSnapping: true });
    handler.setSelectedFeatureIndexes([0]);
    handler._renderSnapEditHandles = true;
    expect(handler.shouldRenderSnapHandles()).toBeTruthy();
  });

  test('getEditHandles()', () => {
    handler.setModeConfig({ enablePolygonSnapping: true });
    handler.setSelectedFeatureIndexes([0]);
    handler._renderSnapEditHandles = true;
    handler._selectedSnapPoint = turfPoint([1, 2]);
    handler._nonSelectedSnapPoint = turfPoint([2, 3]);
    const editHandles = handler.getEditHandles();
    expect(editHandles.length).toEqual(2);
    expect(editHandles).toMatchSnapshot();
  });

  test('getNearestPolygonIndexes()', () => {
    const numberToTrack = 3;
    const nearestPolygons = handler.getNearestPolygonIndexes({ numberToTrack });
    expect(nearestPolygons.length).toEqual(numberToTrack);
    expect(nearestPolygons).toEqual([0, 3, 4]);
  });

  test('isSinglePolygonSelected() - many polygons selected', () => {
    handler.setSelectedFeatureIndexes([0, 1]);
    expect(handler.isSinglePolygonSelected()).toBeFalsy();
  });

  test('isSinglePolygonSelected() - point selected', () => {
    handler.setSelectedFeatureIndexes([2]);
    expect(handler.isSinglePolygonSelected()).toBeFalsy();
  });

  test('isSinglePolygonSelected() - no feature selected', () => {
    handler.setSelectedFeatureIndexes([]);
    expect(handler.isSinglePolygonSelected()).toBeFalsy();
  });

  test('isSinglePolygonSelected() - one polygon selected', () => {
    expect(handler.isSinglePolygonSelected()).toBeTruthy();
  });

  test('getSnapStrengthModifier()', () => {
    handler.setSelectedFeatureIndexes([0]);
    const snapStrengthModifier = handler.getSnapStrengthModifier();
    expect(snapStrengthModifier).toMatchSnapshot();
  });

  test('hasSelectedBeenSnapped() - positive case', () => {
    const [selectedIndex] = handler.getSelectedFeatureIndexes();
    handler.cacheSnapAssociates(selectedIndex, 5);
    expect(handler.hasSelectedBeenSnapped()).toBeTruthy();
  });

  test('hasSelectedBeenSnapped() - negative case', () => {
    expect(handler.hasSelectedBeenSnapped()).toBeFalsy();
  });

  test('shouldPerformSnap() - positive case', () => {
    handler.setModeConfig({ enablePolygonSnapping: true });
    expect(handler.shouldPerformSnap()).toBeTruthy();
  });

  test('shouldPerformSnap() - many selected polygons', () => {
    handler.setSelectedFeatureIndexes([0, 1]);
    handler.setModeConfig({ enablePolygonSnapping: true });
    expect(handler.shouldPerformSnap()).toBeFalsy();
  });

  test('shouldPerformSnap() - no selected features', () => {
    handler.setSelectedFeatureIndexes([]);
    handler.setModeConfig({ enablePolygonSnapping: true });
    expect(handler.shouldPerformSnap()).toBeFalsy();
  });

  test('shouldPerformSnap() - selected point', () => {
    handler.setSelectedFeatureIndexes([2]);
    handler.setModeConfig({ enablePolygonSnapping: true });
    expect(handler.shouldPerformSnap()).toBeFalsy();
  });

  test('shouldPerformSnap() - is already snapped', () => {
    const [selectedIndex] = handler.getSelectedFeatureIndexes();
    handler.cacheSnapAssociates(selectedIndex, 5);
    handler.setModeConfig({ enablePolygonSnapping: true });
    expect(handler.shouldPerformSnap()).toBeFalsy();
  });

  test('shouldPerformSnap() - enablePolygonSnapping is false', () => {
    handler.setModeConfig({ enablePolygonSnapping: false });
    expect(handler.shouldPerformSnap()).toBeFalsy();
  });

  test('shouldPerformUnsnap() - positive case', () => {
    const [selectedIndex] = handler.getSelectedFeatureIndexes();
    handler.cacheSnapAssociates(selectedIndex, 5);
    handler.setModeConfig({ enablePolygonSnapping: true });
    expect(handler.shouldPerformUnsnap()).toBeTruthy();
  });

  test('shouldPerformUnsnap() - many polygons selected', () => {
    handler.setSelectedFeatureIndexes([0, 1]);
    handler.setModeConfig({ enablePolygonSnapping: true });
    expect(handler.shouldPerformUnsnap()).toBeFalsy();
  });

  test('shouldPerformUnsnap() - no selected features', () => {
    handler.setSelectedFeatureIndexes([]);
    handler.setModeConfig({ enablePolygonSnapping: true });
    expect(handler.shouldPerformUnsnap()).toBeFalsy();
  });

  test('shouldPerformUnsnap() - selected point', () => {
    handler.setSelectedFeatureIndexes([2]);
    handler.setModeConfig({ enablePolygonSnapping: true });
    expect(handler.shouldPerformUnsnap()).toBeFalsy();
  });

  test('shouldPerformUnsnap() - is not already snapped', () => {
    handler.setModeConfig({ enablePolygonSnapping: true });
    expect(handler.shouldPerformUnsnap()).toBeFalsy();
  });

  test('shouldPerformUnsnap() - enablePolygonSnapping is false', () => {
    handler.setModeConfig({ enablePolygonSnapping: false });
    expect(handler.shouldPerformSnap()).toBeFalsy();
  });

  test('shouldPerformStandardModeAction() - selected has not been snapped', () => {
    expect(handler.shouldPerformStandardModeAction()).toBeTruthy();
  });

  test('shouldPerformStandardModeAction() - multiple polygons selected', () => {
    handler.setSelectedFeatureIndexes([0, 1]);
    expect(handler.shouldPerformStandardModeAction()).toBeTruthy();
  });

  test('shouldPerformSnap() - selected point', () => {
    handler.setSelectedFeatureIndexes([2]);
    expect(handler.shouldPerformStandardModeAction()).toBeTruthy();
  });

  test('shouldPerformSnap() - selected polygon as been snapped', () => {
    const [selectedIndex] = handler.getSelectedFeatureIndexes();
    handler.cacheSnapAssociates(selectedIndex, 5);
    expect(handler.shouldPerformStandardModeAction()).toBeFalsy();
  });

  test('getSnapDetailsFromCandidates()', () => {
    const snapDetails = handler.getSnapDetailsFromCandidates([0, 3, 4]);
    expect(snapDetails).toMatchSnapshot();
  });

  test('cacheSnapAssociates() - one snap association', () => {
    handler.cacheSnapAssociates(5, 8);
    expect(handler._snapAssociations[5]).toEqual([8]);
    expect(handler._snapAssociations[8]).toEqual([5]);
  });

  test('cacheSnapAssociates() - multiple snap associations', () => {
    handler.cacheSnapAssociates(5, 8);
    handler.cacheSnapAssociates(5, 9);
    expect(handler._snapAssociations[5]).toEqual([8, 9]);
    expect(handler._snapAssociations[8]).toEqual([5]);
    expect(handler._snapAssociations[9]).toEqual([5]);
  });

  test('getSnapAssociates()', () => {
    handler.cacheSnapAssociates(5, 8);
    handler.cacheSnapAssociates(5, 9);
    expect(handler.getSnapAssociates(5)).toEqual([8, 9]);
  });

  test('clearSnapAssociates()', () => {
    handler.cacheSnapAssociates(5, 8);
    handler.cacheSnapAssociates(5, 9);
    expect(handler.getSnapAssociates(5)).toEqual([8, 9]);
    handler.clearSnapAssociates(8);
    expect(handler._snapAssociations[5]).toEqual([9]);
    expect(handler._snapAssociations[9]).toEqual([5]);
    expect(handler._snapAssociations[8]).toEqual([]);
  });

  test('calculateSnapMove() - polygon distance within threshold', () => {
    const snapDetails = {
      index: 0,
      nonSelectedSnapPoint: {
        geometry: {
          coordinates: [-122.4425271525049, 37.96332765658352],
          type: 'Point'
        },
        properties: {},
        type: 'Feature'
      },
      selectedSnapEdgeLength: 2.4850484845856666,
      selectedSnapPoint: {
        geometry: {
          coordinates: [-122.40815000473599, 37.964108978087694],
          type: 'Point'
        },
        properties: {},
        type: 'Feature'
      },
      snapDistance: 0.1
    };

    const snapMove = handler.calculateSnapMove(snapDetails, 0.1);
    const originalPolygon = polygonsForSnappingTests.features[snapDetails.index];
    expect(snapMove).toBeDefined();
    if (snapMove) {
      expect(originalPolygon).not.toEqual(snapMove.movedPolygon);
      expect(snapMove).toMatchSnapshot();
    }
  });

  test('calculateSnapMove() - polygon distance exceeds threshold', () => {
    const snapDetails = {
      index: 0,
      nonSelectedSnapPoint: {
        geometry: {
          coordinates: [-122.4425271525049, 37.96332765658352],
          type: 'Point'
        },
        properties: {},
        type: 'Feature'
      },
      selectedSnapEdgeLength: 2.4850484845856666,
      selectedSnapPoint: {
        geometry: {
          coordinates: [-122.40815000473599, 37.964108978087694],
          type: 'Point'
        },
        properties: {},
        type: 'Feature'
      },
      snapDistance: 5
    };

    const snapMove = handler.calculateSnapMove(snapDetails, 0.1);
    expect(snapMove).toBeNull();
  });
});
