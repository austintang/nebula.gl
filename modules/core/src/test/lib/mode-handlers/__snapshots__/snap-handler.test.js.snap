// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`polygon snapping getEditHandles() 1`] = `
Array [
  Object {
    "featureIndex": -1,
    "position": Array [
      1,
      2,
    ],
    "positionIndexes": Array [
      -1,
    ],
    "type": "intermediate",
  },
  Object {
    "featureIndex": -1,
    "position": Array [
      2,
      3,
    ],
    "positionIndexes": Array [
      -1,
    ],
    "type": "intermediate",
  },
]
`;

exports[`polygon snapping getSnapDetailsFromCandidates() 1`] = `
Object {
  "index": 0,
  "nonSelectedSnapPoint": Object {
    "geometry": Object {
      "coordinates": Array [
        -122.44252582524939,
        37.987923255302974,
      ],
      "type": "Point",
    },
    "properties": 1,
    "type": "Feature",
  },
  "selectedSnapPoint": Object {
    "geometry": Object {
      "coordinates": Array [
        -122.40814940182412,
        37.97528325161274,
      ],
      "type": "Point",
    },
    "properties": Object {},
    "type": "Feature",
  },
  "snapDistance": 3.324623681425176,
}
`;

exports[`polygon snapping getSnapStrengthModifier() 1`] = `1.069575674579367`;
